code-dot-org/code-dot-org
# Localize the password-reset email

[FND-1481](https://codedotorg.atlassian.net/browse/FND-1481)
Localize the password-reset email, which is sent by `Devise::Mailer`. Strings in this email are very sensitive. Thus, they are entrusted to only approved translators in the **Code.org - Restricted** CrowdIn project.  

**This PR contains:**
* Localized password-reset email template `dashboard/app/views/devise/mailer/reset_password_instructions.html.erb`.
* English strings from the password-reset email `dashboard/config/locales/restricted.en.yml`.
* Configuration for the sync-in step `bin/i18n-codeorg/in.sh`.
* Configuration for the sync-up step `bin/i18n/i18n_script_utils.rb`, `bin/i18n/codeorg_restricted_crowdin.yml`, and `bin/i18n/codeorg_crowdin.yml`.
* Unit test in `dashboard/test/models/user_test.rb`.

## Links
* [Code.org - Restricted](https://crowdin.com/project/codeorg-restricted/settings#members) CrowdIn project

## Testing story
1. Run the full I18n pipeline sync-in, sync-up, sync-down, and sync-out locally.
2. Test reseting password and verify that emails sent are localized. 
  * Use `mailcatcher` to intercept emails sent on the local machine.
  * For regular account (user who doesn't have associated child accounts)

    *  Password-reset email in English
        <img width="654" alt="Screen Shot 2021-07-20 at 3 08 04 AM" src="https://user-images.githubusercontent.com/46507039/126221280-59f55f42-910f-4de5-8b27-ddb11b2fa42e.png">

    *  Password-reset email in es-MX
        <img width="722" alt="Screen Shot 2021-07-21 at 12 05 43 AM" src="https://user-images.githubusercontent.com/46507039/126366027-f239f8bf-c520-4f29-9fcf-0cac79612fbd.png">


  * For user with associated child accounts
    *  Password-reset email in English
        <img width="747" alt="Screen Shot 2021-07-20 at 3 07 40 AM" src="https://user-images.githubusercontent.com/46507039/126221276-2d3b5cac-9be7-4db4-b801-e16dc29bdb82.png">

    *  Password-reset email in es-MX
        <img width="574" alt="Screen Shot 2021-07-21 at 12 05 53 AM" src="https://user-images.githubusercontent.com/46507039/126366076-0ff401b6-f5f5-48f6-b9ac-14e57efbe30e.png">


3. Run unit tests: `bundle exec rails test test/models/user_test.rb`

## Follow-up work
- [x] Make the sync-up step to upload `i18n/locales/source/dashboard/restricted.yml` to `dasboard/restricted.yml` in Crowdin.
- [x] Make the sync-down step to download translated `dasboard/restricted.yml` files to `i18n/locales/<locale_code>/dashboard/restricted.yml`. Currently, it downloads translated files to `i18n/locales/<language_name>/dashboard/restricted.yml`. See [this line](https://github.com/code-dot-org/code-dot-org/blob/6c0f894a2e1235cea90c6231f57f677f9ef11cdf/lib/cdo/crowdin/utils.rb#L93).
- [x] Delete `subject` string in `reset_password_instructions` in [`devise.en.yml`](https://github.com/code-dot-org/code-dot-org/blob/49f6d4b2f1c3c85c799eba1ffb60e2b437bd652c/dashboard/config/locales/devise.en.yml#L23)
- [x] Add `bin/i18n/codeorg_restricted_credentials.yml` file to the i18n-dev server.
- [x] Delete test translations in `Code.org - Restricted` project.
- [x] Copy the list of languages from `Code.org` project to `Code.org - Restricted` project. ([link](https://crowdin.com/project/codeorg-restricted/settings#translations))


## Security
Only a few approved translators, likely all Code.org members, can join the **Code.org - Restricted** project.

## PR Checklist:
